Descriptor cluser's new attribute : EndpointUniqueId is added as per latest spec 0.7-fall2024-ncr(V1.5)

suneelsignify opened the pull request:
Specification for V1.5 (https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/9171) defines a new attribute EndpointUniqueId in Descriptor cluster.

**Feature explanation:** 
One of the TCR-features in 1.5 is the addition of the (readonly) EndpointUniqueID attribute in the Descriptor cluster. This is filled by the server with a unique value and allows recognize/deduplicate when a client encounters a device via multiple paths.
The (already existing) UniqueID attribute in Basic Information (and Bridged Device Basic Information) cluster can be used to detect multiple paths between a device and a client.
The EndpointUniqueID (this PR) provides a similar mechanism for functionality on endpoints within those devices.
PPT: `https://docs.google.com/presentation/d/1r2BBtPPyOGDni3CBNa9sD90ORMdaA2owqjIwcSRh8K0`
Spec changes: `https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/9171`
This PR also adds the test for the new attribute ( test plan changes are in `https://github.com/CHIP-Specifications/chip-test-plans/pull/4487` )

**General Changes:**
Descriptor cluster implementation  : Changed to implement  the new attribute support.
 Data model provider : To access the ember bindings
attribute storage : To define/ store and access the attribute


### Testing

1. Automated python test TC_DESC_2_1 is added which verifies the EndpointUniqueID.
2. Also it was verified using the bridge-app as the DUT device by reading the EndpointUniqueId attribute using the chip-tool. 

**Following are the manual test details:**

Applications used:
1. Test DUT Application used : bridge-app (linux-x64-bridge)
2. Chip tool used : linux-x64-chip-tool

Test command used:
1. `chip-tool pairing onnetwork 1 20202021`
2. `chip-tool descriptor read endpoint-unique-id 1 7`

Expected result : 
` EndpointUniqueID: AABBCCDDEEFFGGHHIIJJKKLLMMNNOO02`

Actual result:
```
[1744721452.752] [436926:436928] [TOO] Endpoint: 8 Cluster: 0x0000_001D Attribute 0x0000_0005 DataVersion: 1384225220
[1744721452.752] [436926:436928] [TOO]   EndpointUniqueID: AABBCCDDEEFFGGHHIIJJKKLLMMNNOO02
[1744721452.752] [436926:436928] [EM] <<< [E:24153i S:23121 M:206193812 (Ack:246835769)] (S) Msg TX from 000000000001B669 to 1:0000000000000001 [B7F5] [UDP:[fe80::e6db:bf5:2882:6c4f%wlp0s20f3]:5540] --- Type 0000:10 (SecureChannel:StandaloneAck) (B:34)
```

suneelsignify committed:
Descriptor cluser's new attribute : EndpointUniqueId is added as per latest spec 0.7-fall2024-ncr

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

suneelsignify committed:
Added exception to allow #include<string> in af-types.h

suneelsignify committed:
Merge branch 'master' into master

jmartinez-silabs commented on src/app/clusters/descriptor/descriptor.cpp:
```diff
@@ -211,6 +212,18 @@ CHIP_ERROR DescriptorAttrAccess::ReadDeviceAttribute(EndpointId endpoint, Attrib
+
+CHIP_ERROR DescriptorAttrAccess::ReadEndpointUniqueId(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    char endpointUniqueId[32 + 1] = { 0 };
```
can we get a consexpr for this 32 with a name that reflect what it is and why it is 32

suneelsignify replied:
Accepted, changed to chip::app::Clusters::Descriptor::Attributes::EndpointUniqueId::TypeInfo::MaxLength()

jmartinez-silabs commented on src/app/util/attribute-storage.cpp:
```diff
@@ -1080,6 +1080,18 @@ CHIP_ERROR GetSemanticTagForEndpointAtIndex(EndpointId endpoint, size_t index,
+{
+    uint16_t endpointIndex = emberAfIndexFromEndpoint(endpoint);
+
+    if (endpointIndex == 0xFFFF)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
```
```suggestion
    VerifyOrReturnError(endpointIndex != chip::kInvalidEndpointId, CHIP_ERROR_NOT_FOUND);
```

suneelsignify replied:
Agree, the current implementation is fine.

jmartinez-silabs commented on src/app/util/attribute-storage.cpp:
```diff
@@ -1104,6 +1116,18 @@ CHIP_ERROR SetTagList(EndpointId endpoint, Span<const Clusters::Descriptor::Stru
+CHIP_ERROR SetEndPointUniqueId(chip::EndpointId endpoint, std::string endpointUniqueId)
+{
+    uint16_t endpointIndex = emberAfIndexFromEndpoint(endpoint);
+    if (endpointIndex == 0xFFFF)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
```
```suggestion
    VerifyOrReturnError(endpointIndex != chip::kInvalidEndpointId, CHIP_ERROR_INVALID_ARGUMENT);
```

suneelsignify replied:
As per the above comment : "endpoint indices are not the same as endpoint ids." keeping the current implementation.

bzbarsky-apple commented on src/app/util/af-types.h:
```diff
@@ -236,6 +237,8 @@ struct EmberAfDefinedEndpoint
      */
     chip::Span<const chip::app::Clusters::Descriptor::Structs::SemanticTagStruct::Type> tagList;
+
+    std::string endpointUniqueId;
```
Not OK to use std::string in general in embedded code.  This needs a different representation.

suneelsignify replied:
Agreed. Changed to char array.

bzbarsky-apple commented on src/app/util/attribute-storage.cpp:
```diff
@@ -1104,6 +1116,18 @@ CHIP_ERROR SetTagList(EndpointId endpoint, Span<const Clusters::Descriptor::Stru
 
+CHIP_ERROR SetEndPointUniqueId(chip::EndpointId endpoint, std::string endpointUniqueId)
+{
+    uint16_t endpointIndex = emberAfIndexFromEndpoint(endpoint);
+    if (endpointIndex == 0xFFFF)
```
This looks wrong to me.  Shouldn't it be possible to set the unique ID of a disabled endpoint before enabling it?

But separately: are unique IDs allowed to change?  If not, shouldn't the unique ID be part of the "add the endpoint" machinery, not a separate setter?

I know you copy/pasted this stuff from the existing SetTagList bits and whatnot, but those look wrong too.

suneelsignify replied:
Agreed, removed the setter and now it is part of emberAfSetDynamicEndpoint.

suneelsignify committed:
Merge branch 'master' into master

suneelsignify committed:
According to the review comments:
1. Removed the hardcoaded size of the array to store endpointUniqueId
2. Changed the way endpointUniqueId is set. Using the add end point engine rather than a new setter.
3. Not using std::string instead changed to char array.

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

suneelsignify committed:
Changing bridge app to support the new arrtibute endpointUniqueId

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

suneelsignify committed:
Merge branch 'master' into master

suneelsignify committed:
Merge branch 'master' into master

suneelsignify committed:
Adding support for EndpointUniqueId in examples.

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

suneelsignify committed:
Compiler error, typo fixed.

suneelsignify committed:
Compiler error : typo fixed

restyled-commits committed:
Restyled by clang-format

suneelsignify committed:
Compiler error : typo fixed

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

suneelsignify committed:
Compiler error : Typo fixed.

suneelsignify committed:
Added EndpointUniqueId support in subdevice example.

suneelsignify committed:
Fixed compiler error

suneelsignify committed:
Merge branch 'master' into master

s07641069 reviewed (approved):
Telink related changes looks ok

suneelsignify committed:
Merge branch 'master' into master

suneelsignify committed:
Merge branch 'master' into master

andy31415 reviewed (dismissed):
Since testing part is manual, please expand beyond `Testing was done using the bridge-app and the DUT deive by reading the EndpointUniqueId attribute using the chip-tool.`:

- [ ] what commands exactly were run and what was the observed output (i.e. chip-tool commands)
- [ ] explain why automating this test is not possible at this time

We want to encourage people to write automated tests and avoid having "manually tested" because it bypasses extra work to automate tests (i.e. use manual tests only when automation is not possible ... and then have a list of why automation was not possible and hopefully some day remove those roadblocks).

suneelsignify committed:
Merge branch 'project-chip:master' into master

suneelsignify committed:
Adding zap generated for the newly added examples

bzbarsky-apple commented on src/app/clusters/descriptor/descriptor.cpp:
```diff
@@ -211,6 +212,15 @@ CHIP_ERROR DescriptorAttrAccess::ReadDeviceAttribute(EndpointId endpoint, Attrib
 
+CHIP_ERROR DescriptorAttrAccess::ReadEndpointUniqueId(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    char epUniqueId[Attributes::EndpointUniqueId::TypeInfo::MaxLength() + 1] = { 0 };
```
Why +1 ?

suneelsignify replied:
Yes, +1 not needed.

bzbarsky-apple commented on src/app/util/attribute-storage.cpp:
```diff
@@ -265,7 +265,7 @@ uint16_t emberAfGetDynamicIndexFromEndpoint(EndpointId id)
 CHIP_ERROR emberAfSetDynamicEndpoint(uint16_t index, EndpointId id, const EmberAfEndpointType * ep,
                                      const Span<DataVersion> & dataVersionStorage, Span<const EmberAfDeviceType> deviceTypeList,
-                                     EndpointId parentEndpointId)
+                                     chip::CharSpan endpointUniqueId, EndpointId parentEndpointId)
```
A few things:

1.  This is an API change, which is why you had to change a whole bunch of other code.  What we should probably do is keep both APIs (i.e. add a second function signature that has the unique ID, internally make the version that does not take that call the one that does an pass an empty CharSpan.
2. Please take out the unnecessary `chip::` prefixes throughout.

bzbarsky-apple commented on src/app/util/attribute-storage.cpp:
```diff
@@ -321,6 +321,11 @@ CHIP_ERROR emberAfSetDynamicEndpoint(uint16_t index, EndpointId id, const EmberA
     emAfEndpoints[index].deviceTypeList = deviceTypeList;
     emAfEndpoints[index].endpointType   = ep;
     emAfEndpoints[index].dataVersions   = dataVersionStorage.data();
+    memcpy((void *) (emAfEndpoints[index].endpointUniqueId), endpointUniqueId.data(),
+           endpointUniqueId.size() > Clusters::Descriptor::Attributes::EndpointUniqueId::TypeInfo::MaxLength()
+               ? Clusters::Descriptor::Attributes::EndpointUniqueId::TypeInfo::MaxLength()
+               : endpointUniqueId.size());
```
A suggestion to not reinvent the wheel:
```
MutableCharSpan targetSpan(emAfEndpoints[index].endpointUniqueId);
CopyCharSpanToMutableCharSpanWithTruncation(endpointUniqueId, targetSpan);
```

suneelsignify replied:
Agree!

bzbarsky-apple commented on src/app/util/attribute-storage.cpp:
```diff
@@ -1080,6 +1085,22 @@ CHIP_ERROR GetSemanticTagForEndpointAtIndex(EndpointId endpoint, size_t index,
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    chip::CharSpan epUniqueIdSpan(emAfEndpoints[endpointIndex].endpointUniqueId,
+                                  strnlen(emAfEndpoints[endpointIndex].endpointUniqueId,
```
This is broken: nothing guarantees null-termination for these things.  You need to store the length of the data separately.

suneelsignify replied:
Added size.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/descriptor-cluster.xml:
```diff
@@ -62,5 +62,8 @@ limitations under the License.
       </mandatoryConform>
     </attribute>
+    <attribute side="server" code="0x0005" name="EndpointUniqueId" define="END_POINT_UNIQUE_ID" type="char_string" optional="true" length="32">
+      <mandatoryConform/>        
```
I am curious how this XML was generated, because it's not making sense.

suneelsignify replied:
I manually added this line and ran zap_regen_all.py which resulted a lot of auto generated code. Please let me know how this xml can be generated. (Doing it for the first time)

suneelsignify replied:
Now generated using alchemy

bzbarsky-apple commented on src/app/zap-templates/zcl/zcl-with-test-extensions.json:
```diff
@@ -203,7 +203,7 @@
         "Chime": ["SelectedChime", "Enabled"],
         "Closure Control": ["OverallState", "OverallTarget"],
-        "Descriptor": ["ClusterRevision", "FeatureMap"],
+        "Descriptor": ["ClusterRevision", "FeatureMap", "EndpointUniqueId"],
```
This isn't using the right attribute name....  Did you check that this part actually works, by looking at generated endpoint_config?

suneelsignify replied:
I assumed zap_regen_all.py will take of that. Can you please let me how to generate endpoint_config ?

bzbarsky-apple commented on src/app/zap-templates/zcl/zcl.json:
```diff
@@ -197,7 +197,7 @@
         "Chime": ["SelectedChime", "Enabled"],
         "Closure Control": ["OverallState", "OverallTarget"],
-        "Descriptor": ["ClusterRevision", "FeatureMap"],
+        "Descriptor": ["ClusterRevision", "FeatureMap", "EndpointUniqueId"],
```
See above.

suneelsignify committed:
Merge branch 'master' into master

suneelsignify committed:
changes made according to the review comments

restyled-commits committed:
Restyled by whitespace

suneelsignify committed:
Merge branch 'master' into master

suneelsignify committed:
Generated descriptor-cluster.xml using alchemy and generated zap from it

restyled-commits committed:
Restyled by clang-format

suneelsignify committed:
Adding a YAML test to check EndpointUniqueID

bzbarsky-apple commented on examples/bridge-app/asr/subdevice/SubDeviceManager.cpp:
Why the random mode changes for this file and others?

suneelsignify replied:
Looks like something changed the mode of these files. Fixed it.

bzbarsky-apple commented on examples/bridge-app/linux/main.cpp:
```diff
@@ -121,6 +121,7 @@ DECLARE_DYNAMIC_ATTRIBUTE(Descriptor::Attributes::DeviceTypeList::Id, ARRAY, kDe
     DECLARE_DYNAMIC_ATTRIBUTE(Descriptor::Attributes::ServerList::Id, ARRAY, kDescriptorAttributeArraySize, 0), /* server list */
     DECLARE_DYNAMIC_ATTRIBUTE(Descriptor::Attributes::ClientList::Id, ARRAY, kDescriptorAttributeArraySize, 0), /* client list */
     DECLARE_DYNAMIC_ATTRIBUTE(Descriptor::Attributes::PartsList::Id, ARRAY, kDescriptorAttributeArraySize, 0),  /* parts list */
+    DECLARE_DYNAMIC_ATTRIBUTE(Descriptor::Attributes::EndpointUniqueID::Id, ARRAY, 32, 0),                      /* parts list */
```
Wrong comment.

bzbarsky-apple commented on examples/bridge-app/linux/main.cpp:
Not sure why any changes are needed to this file at all....  But OK.

suneelsignify replied:
To test the feature added support of EndPointUniueID .

bzbarsky-apple commented on examples/bridge-app/telink/src/AppTask.cpp:
Again, why mode change?

bzbarsky-apple commented on examples/fabric-bridge-app/fabric-bridge-common/include/BridgedDeviceManager.h:
Why all these mode changes?  Please look through the diff and fix all of these.

bzbarsky-apple commented on src/app/util/attribute-storage.cpp:
```diff
@@ -321,6 +329,11 @@ CHIP_ERROR emberAfSetDynamicEndpoint(uint16_t index, EndpointId id, const EmberA
+
+    MutableCharSpan targetSpan(emAfEndpoints[index].endpointUniqueId);
+    CopyCharSpanToMutableCharSpanWithTruncation(endpointUniqueId, targetSpan);
+    emAfEndpoints[index].endpointUniqueIdSize = endpointUniqueId.size();
```
This is a buffer overrun right here.  Please figure out why and fix it.

suneelsignify replied:
should be targetSpan.size()

bzbarsky-apple commented on src/app/util/attribute-storage.cpp:
```diff
@@ -1080,6 +1093,21 @@ CHIP_ERROR GetSemanticTagForEndpointAtIndex(EndpointId endpoint, size_t index,
+    }
+
+    CharSpan epUniqueIdSpan(emAfEndpoints[endpointIndex].endpointUniqueId, emAfEndpoints[endpointIndex].endpointUniqueIdSize);
+    CopyCharSpanToMutableCharSpan(epUniqueIdSpan, epUniqueIdMutSpan);
```
```suggestion
    return CopyCharSpanToMutableCharSpan(epUniqueIdSpan, epUniqueIdMutSpan);
```

Otherwise this happily returns CHIP_NO_ERROR on failure to copy!

suneelsignify replied:
Agree.

suneelsignify committed:
Adhering to the review comments

restyled-commits committed:
Restyled by clang-format

suneelsignify committed:
Added python test to verify TC_DESC_2_2 test case

restyled-commits committed:
Restyled by autopep8

suneelsignify committed:
Clean up in test script.

restyled-commits committed:
Restyled by autopep8

suneelsignify commented:
> Since testing part is manual, please expand beyond `Testing was done using the bridge-app and the DUT deive by reading the EndpointUniqueId attribute using the chip-tool.`:
> 
> * [ ]  what commands exactly were run and what was the observed output (i.e. chip-tool commands)
> * [ ]  explain why automating this test is not possible at this time
> 
> We want to encourage people to write automated tests and avoid having "manually tested" because it bypasses extra work to automate tests (i.e. use manual tests only when automation is not possible ... and then have a list of why automation was not possible and hopefully some day remove those roadblocks).

We added the automated tests now python test src/python_testing/TC_DeviceBasicComposition.py

suneelsignify committed:
Fix CI error

andy31415 commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -941,6 +941,217 @@ def log_structured_data(start_tag: str, dump_string):
+        if len(parts_list_ep_0) == 0:
+            self.fail_current_test()
+        self.print_step("1a.5", "Verify that Revision should match the revision of that Device Type.")
+        if device_revision_ep_0[0] != 1:
```
This hard-codes the revision ... should this not be somehow from the spec XMLs so it is dynamic based on spec?

andy31415 commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -941,6 +941,217 @@ def log_structured_data(start_tag: str, dump_string):
+                    if item not in [14, 256, 19, 770, 17, 18, 20]:
+                        self.fail_current_test()
+
+                if cluster_revision_ep_non_zero <= 1:
```
please use constants instead of magic values

andy31415 replied:
Re-opening: we still have `1` as a magic number. Why is 1 not a valid revision number?

andy31415 commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -941,6 +941,217 @@ def log_structured_data(start_tag: str, dump_string):
+                self.print_step(
+                    "1b.1", "Verify that the DeviceTypeList count is at least one for end point {endpoint_id}")
+
+                if len(listed_device_types_ep_non_zero) < 1:
```
it is more pythonic to do:

```suggestion
                if not listed_device_types_ep_non_zero:
```

andy31415 commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -941,6 +941,217 @@ def log_structured_data(start_tag: str, dump_string):
+                device_revision_ep_non_zero = [i.revision for i in endpoint[Clusters.Descriptor]
+                                               [Clusters.Descriptor.Attributes.DeviceTypeList]]
+                for device_revision in device_revision_ep_non_zero:
+                    if device_revision != 1:
```
replace all magic values with constants.

For specific revision values, we should read that from spec XML.

suneelsignify commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -941,6 +941,217 @@ def log_structured_data(start_tag: str, dump_string):
+        if device_revision_ep_0[0] != 1:
+            self.fail_current_test()
+        self.print_step("1a.5", "Verify that cluster Revision should not be less than 1")
+        if cluster_revision < 1:
```
According to the test spec (https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/cluster/Descriptor.adoc) step 1.a  (Revision value is not less than 1.) which is checked in the script.

suneelsignify committed:
Changes according to review comments.

andy31415 commented on src/app/tests/suites/TestDescriptorClusterEndpointUniqueID.yaml:
```diff
@@ -0,0 +1,35 @@
+      command: "readAttribute"
+      attribute: "EndpointUniqueId"
+      response:
+          value: "AABBCCDDEEFFGGHHIIJJKKLLMMNNOO01"
```
why is this the correct unique ID for general devices? This seems to be example-specific.

suneelsignify replied:
Yes, it will be different for each device. Hence we can test it in Python tests, removing the yaml tests.

suneelsignify committed:
Removing the yaml file as python test is added for testing EndpointUniqueID

bzbarsky-apple commented on src/app/clusters/descriptor/descriptor.cpp:
Why the mode change?

bzbarsky-apple commented on src/app/util/af-types.h:
Again: why the mode change?

Please see https://github.com/project-chip/connectedhomeip/pull/37990#discussion_r2021844017

bzbarsky-apple commented on src/app/util/af-types.h:
```diff
@@ -236,6 +236,13 @@ struct EmberAfDefinedEndpoint
+    /**
+     * Unique Id for this end point.
+     */
+    char endpointUniqueId[chip::app::Clusters::Descriptor::Attributes::EndpointUniqueID::TypeInfo::MaxLength()] = { 0 };
```
So as you might have noticed... this has RAM costs, since we are allocating this storage unconditionally, even for applications that don't want to implement the new optional attribute.  Should this be conditioned on some build flag?

Alternately, you could use a setup similar to TagList and have minimal storage costs here (basically two words) and the rest of the cost is paid by the application, if it decides to provide this data.

bzbarsky-apple commented on src/app/util/attribute-storage.h:
```diff
@@ -259,6 +261,35 @@ CHIP_ERROR emberAfSetDynamicEndpoint(uint16_t index, chip::EndpointId id, const
+CHIP_ERROR emberAfSetDynamicEndpointWithEpUniqueId(uint16_t index, chip::EndpointId id, const EmberAfEndpointType * ep,
+                                                   const chip::Span<chip::DataVersion> & dataVersionStorage,
+                                                   chip::Span<const EmberAfDeviceType> deviceTypeList = {},
+                                                   chip::CharSpan endpointUniqueId                    = {},
```
This argument needs lifetime documentation.  With the current implementation, we would document that the data will be copied out of endpointUniqueId and it does not need to survive once this call returns.  If we change to just referencing the data, it would be documented similar to the deviceTypeList bits above.

suneelsignify committed:
Changes made compling with review comments

andy31415 reviewed (changes requested):
This plumbs in direct ember bindings only for this via compile-time configurations that are application specific.

This needs at a minimal plumbing through the provider vtable at https://github.com/project-chip/connectedhomeip/blob/master/src/app/data-model-provider/MetadataTypes.h#L46

andy31415 commented on src/app/util/af-types.h:
```diff
@@ -236,6 +236,15 @@ struct EmberAfDefinedEndpoint
+     */
+    char endpointUniqueId[chip::app::Clusters::Descriptor::Attributes::EndpointUniqueID::TypeInfo::MaxLength()] = { 0 };
+
+    size_t endpointUniqueIdSize = 0;
```
is the unique ID bytes? if so it should be uint8_t. If not bytes, then maybe null-terminated and then we do not need a length.

size_t is probably too much ... is u8 or u16 enough (with an assert on sizeof to make sure it never breaks)?

suneelsignify replied:
I changed it according to this feedback : https://github.com/project-chip/connectedhomeip/pull/37990#discussion_r2017512465

andy31415 replied:
Could we make it null-terminated by allocating size + 1 and then not worry about separating size from data? Otherwise I wonder if people will make mistakes by adding these to logs.

andy31415 commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -941,6 +942,259 @@ def log_structured_data(start_tag: str, dump_string):
+        unique_ids = []
+        ROOT_NODE_DEVICE_TYPE = 0x0016
+        server_conformance = 0x001D
+        OnOff_deviceType = 0x0100
```
the syntax of this feels off... constants are generally `UPPER_SNAKE_CASE`: https://peps.python.org/pep-0008/#constants

andy31415 commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -941,6 +942,259 @@ def log_structured_data(start_tag: str, dump_string):
 
+    @async_test_body
+    async def test_TC_DESC_2_1(self):
+        asserts.assert_true('PIXIT.DESC.AGGREVISION' in self.matter_test_config.global_test_params,
```
Is this typical to ask for a revision number as a parameter? why is this not from the spec?

andy31415 commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -941,6 +942,259 @@ def log_structured_data(start_tag: str, dump_string):
+            "1a.4", "Verify that the DeviceTypeList may only contain other Node Device Types (device types with scope=node, it can be any of the following Power Source, OTA Requestor, OTA Provider) next to the Root Node Device Type.")
+
+        if listed_device_types:
+            if not all(x in {ROOT_NODE_DEVICE_TYPE, OTARequestor_deviceType, Aggregator_deviceType} for x in listed_device_types):
```
this feels hard to read. Can we use some set operations? This seems like a set difference that has to be empty.

andy31415 commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -941,6 +942,259 @@ def log_structured_data(start_tag: str, dump_string):
+            self.print_step(8, "TH reads ServerList attribute for endpoint :{endpoint_id}".format(endpoint_id=endpoint_id))
+
+            if endpoint_id != 0:
+                if Clusters.Descriptor not in endpoint:
```
is this possible? Can we have endpoints without descriptors?

suneelsignify committed:
Adding enberAf prefix to GetEndpointUniqueIdForEndPoint as per review feedback as it is a ember public method

suneelsignify committed:
Added Data model provider support to access EndPointUniqueID according to review comments

suneelsignify committed:
Changing as per review comments.

suneelsignify committed:
Merge remote-tracking branch 'upstream/master'

suneelsignify committed:
Merge branch 'master' into master

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

suneelsignify committed:
Resolved merge conflicts

suneelsignify committed:
Resolving merge conflicts

bzbarsky-apple commented on src/app/clusters/descriptor/descriptor.cpp:
```diff
@@ -211,15 +211,19 @@ CHIP_ERROR DescriptorAttrAccess::ReadDeviceAttribute(EndpointId endpoint, Attrib
     return err;
 }
-
+#ifdef CONFIG_USE_ENDPOINT_UNIQUE_ID
```
The way config defines work normally is:

1. They start with CHIP_CONFIG_
2. They are defined to some default in a config header that is guaranteed to include the config headers that applications would then use to override them
3. They are tested with `#if`.

Concretely here: what config header is this file including that guarantees that it will see this define if it's defined?  With the `#if` approach, the compiler will catch that for you, at least with some warning/error flags we use on some platforms.

bzbarsky-apple commented on src/app/util/af-types.h:
```diff
@@ -235,6 +235,15 @@ struct EmberAfDefinedEndpoint
      */
     chip::Span<const chip::app::Clusters::Descriptor::Structs::SemanticTagStruct::Type> tagList;
+
+#ifdef CONFIG_USE_ENDPOINT_UNIQUE_ID
```
Here we absolutely have to make sure this header includes the right config header.  Otherwise we can have different translation units disagree on the layout of this struct!

bzbarsky-apple commented on src/app/util/attribute-storage.cpp:
```diff
@@ -336,6 +337,98 @@ CHIP_ERROR emberAfSetDynamicEndpoint(uint16_t index, EndpointId id, const EmberA
+    return CHIP_NO_ERROR;
+}
+#else
+    return emberAfSetDynamicEndpointWithEpUniqueId(index, id, ep, dataVersionStorage, deviceTypeList, CharSpan{}, parentEndpointId);
```
This code duplication is really sub-optimal.  I would suggest just having the internal function unconditionally defined and ifdef out the part that uses `endpointUniqueId` in it.

bzbarsky-apple commented on src/app/util/attribute-storage.cpp:
```diff
@@ -347,7 +427,7 @@ CHIP_ERROR emberAfSetDynamicEndpointWithEpUniqueId(uint16_t index, EndpointId id
     MutableCharSpan targetSpan(emAfEndpoints[index].endpointUniqueId);
     CopyCharSpanToMutableCharSpanWithTruncation(endpointUniqueId, targetSpan);
-    emAfEndpoints[index].endpointUniqueIdSize = targetSpan.size();
+    emAfEndpoints[index].endpointUniqueIdSize = static_cast<uint8_t>(std::min<size_t>(targetSpan.size(), UINT8_MAX));
```
Why do we need this min business? The size is not going to be greater than `sizeof(emAfEndpoints[index].endpointUniqueId)`, so we could just static_assert that fits in uint8_t?

bzbarsky-apple commented on src/data-model-providers/codegen/CodegenDataModelProvider.cpp:
```diff
@@ -635,6 +635,25 @@ CHIP_ERROR CodegenDataModelProvider::SemanticTags(EndpointId endpointId, ReadOnl
+    MutableCharSpan copiedSpan(buffer, epUniqueIdSpan.size());
+
+    ReturnErrorOnFailure(builder.EnsureAppendCapacity(1));
+    ReturnErrorOnFailure(builder.Append(copiedSpan));
```
A few questions:

1. Why are we ignoring the return value from `emberAfGetEndpointUniqueIdForEndPoint`?
2. Why are we doing a copy onto the stack and then another onto the heap?  Could we just have the emberAf thing return a span pointing directly at its data and just have the builder reference that span, so we're not doing copies at all?  This might be a question for @andy31415 
3. If we _are_ doing the heap allocation, where is that "+1" in the allocation size coming from?

andy31415 replied:
This should take a `MutableCharSpan &` argument and place the value there. Using an array (readonlybuffers are arrays) and of mutable char spans is odd.

suneelsignify committed:
Complying with the review comments.

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

suneelsignify committed:
Complying to review feedback.

suneelsignify committed:
Merge branch 'master' into master

restyled-commits committed:
Restyled by autopep8

andy31415 commented on src/app/clusters/descriptor/descriptor.cpp:
```diff
@@ -210,6 +211,16 @@ CHIP_ERROR DescriptorAttrAccess::ReadDeviceAttribute(EndpointId endpoint, Attrib
+#if CHIP_CONFIG_USE_ENDPOINT_UNIQUE_ID
+CHIP_ERROR DescriptorAttrAccess::ReadEndpointUniqueId(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    ReadOnlyBufferBuilder<MutableCharSpan> epUniqueId;
```
A buffer should be filled with structure data. This fills data with mutable strings, that seems off:
 - why mutable
 - why is this an allocated buffer? Could we return a pointer to the data?
 - bufferbuilder is for returning ARRAYS of data. Here you need a unique value.
 
 So either the unique ID should be part of the endpoint information (https://github.com/project-chip/connectedhomeip/blob/master/src/app/data-model-provider/MetadataTypes.h#L46 as part of `EndpointEntry`) or you should return a single value
 
You could either have an out parameter like `const char *&pUniqueId` where the value is set to a pointer with a sufficient lifetime OR pass in a MutableCharSpan where you ask the span to be filled (and since we have only 32 bytes we can probably fit it on the stack).

This needs changing:
  - do not use array methods for single values (ReadOnlyBuffer is an array)
  - Do not return MutableCharSpan since they are not mutable
  - There are not checks here to validate [0] is even a valid index (what if the buffer is empty)?

andy31415 commented on src/app/server-cluster/testing/EmptyProvider.h:
```diff
@@ -40,6 +40,9 @@ class EmptyProvider : public app::DataModel::Provider
     CHIP_ERROR ClientClusters(EndpointId endpointId, ReadOnlyBufferBuilder<ClusterId> & builder) override;
     CHIP_ERROR ServerClusters(EndpointId endpointId, ReadOnlyBufferBuilder<app::DataModel::ServerClusterEntry> & builder) override;
+#if CHIP_CONFIG_USE_ENDPOINT_UNIQUE_ID
+    CHIP_ERROR EndpointUniqueID(EndpointId endpointId, ReadOnlyBufferBuilder<MutableCharSpan> & builder) override;
```
```suggestion
    CHIP_ERROR EndpointUniqueID(EndpointId endpointId, MutableCharSpan &out_value) override;
```

seems the correct thing to do and let the caller pass in the buffer to store the value (because we have space for 32 bytes). Otherwise return a const char *, but definitely do not return an array of mutable character spans.

andy31415 commented on src/app/util/mock/attribute-storage.cpp:
```diff
@@ -244,6 +254,26 @@ chip::EndpointId emberAfEndpointFromIndex(uint16_t index)
+    VerifyOrReturnError(epConfig != nullptr, CHIP_ERROR_NOT_FOUND);
+
+    // Ensure the provided span is large enough to hold the unique ID
+    if (epUniqueIdSpan.size() < epConfig->endpointUniqueIdSize)
```
If you use `CopyCharSpanToMutableCharSpan` you do not need to do these checks. This applies throughout: instead of size check + truncate copy (which is essentially the copy above) just use `return CopyCharSpanToMutableCharSpan`

Please make this change everywhere you use this patern.

suneelsignify committed:
Complying with review comments

restyled-commits committed:
Restyled by clang-format

suneelsignify committed:
Merge branch 'master' into master

andy31415 commented on examples/bridge-app/linux/main.cpp:
```diff
@@ -982,6 +1008,40 @@ void ApplicationInit()
+    AddDeviceEndpoint(&ComposedTempSensor1, &bridgedTempSensorEndpoint,
+                      Span<const EmberAfDeviceType>(gComposedTempSensorDeviceTypes),
+                      Span<DataVersion>(gComposedTempSensor1DataVersions),
+                      chip::CharSpan("AABBCCDDEEFFGGHHIIJJKKLLMMNNOO01", strlen("AABBCCDDEEFFGGHHIIJJKKLLMMNNOO01")),
```
use _span here and throughout

andy31415 commented on src/python_testing/matter_testing_infrastructure/chip/testing/pics.py:
```diff
@@ -96,9 +81,32 @@ def read_pics_from_file(path: str) -> dict[str, bool]:
+    mytree = ET.fromstring(contents)
+    for pi in mytree.iter('pixitItem'):
+        name = pi.find('itemNumber').text
+        support = pi.find('support').text
```
is support a bool? The type returned says dictionary with bool values.

andy31415 commented on src/python_testing/matter_testing_infrastructure/chip/testing/pics.py:
```diff
@@ -96,9 +81,32 @@ def read_pics_from_file(path: str) -> dict[str, bool]:
             with open(filename, 'r') as f:
                 contents = f.read()
                 pics_dict.update(parse_pics_xml(contents))
+                print("reading pics ****")
```
This seems like a debug print. Replace with logging module or delete debug prints

suneelsignify committed:
Complying with review comments

suneelsignify committed:
Complying with review comments

suneelsignify committed:
Merge branch 'master' into master

restyled-commits committed:
Restyled by clang-format

suneelsignify committed:
Resolving CI errors

bzbarsky-apple commented on src/app/clusters/descriptor/descriptor.cpp:
```diff
@@ -89,6 +89,7 @@ class DescriptorAttrAccess : public AttributeAccessInterface
     CHIP_ERROR ReadServerClusters(EndpointId endpoint, AttributeValueEncoder & aEncoder);
     CHIP_ERROR ReadClusterRevision(EndpointId endpoint, AttributeValueEncoder & aEncoder);
     CHIP_ERROR ReadFeatureMap(EndpointId endpoint, AttributeValueEncoder & aEncoder);
+    CHIP_ERROR ReadEndpointUniqueId(EndpointId endpoint, AttributeValueEncoder & aEncoder);
```
It's odd to have the declaration unconditional but the implementation conditional.

bzbarsky-apple commented on src/app/util/af-types.h:
```diff
@@ -235,6 +235,15 @@ struct EmberAfDefinedEndpoint
+
+#if CHIP_CONFIG_USE_ENDPOINT_UNIQUE_ID
+    /**
+     * Unique Id for this end point.
```
```suggestion
     * Unique Id for this endpoint.
```

suneelsignify committed:
Resolving CI errors

bzbarsky-apple commented on src/app/util/attribute-storage.cpp:
```diff
@@ -336,6 +344,11 @@ CHIP_ERROR emberAfSetDynamicEndpoint(uint16_t index, EndpointId id, const EmberA
+#if CHIP_CONFIG_USE_ENDPOINT_UNIQUE_ID
+    MutableCharSpan targetSpan(emAfEndpoints[index].endpointUniqueId);
+    CopyCharSpanToMutableCharSpanWithTruncation(endpointUniqueId, targetSpan);
+    emAfEndpoints[index].endpointUniqueIdSize = static_cast<uint8_t>(targetSpan.size());
```
Should static_assert that `sizeof(emAfEndpoints[index].endpointUniqueId) <= UINT8_MAX`, right?

bzbarsky-apple commented on src/app/util/attribute-storage.h:
```diff
@@ -208,6 +208,8 @@ CHIP_ERROR GetSemanticTagForEndpointAtIndex(chip::EndpointId endpoint, size_t in
 CHIP_ERROR SetTagList(chip::EndpointId endpoint,
                       chip::Span<const chip::app::Clusters::Descriptor::Structs::SemanticTagStruct::Type> tagList);
 
+CHIP_ERROR emberAfGetEndpointUniqueIdForEndPoint(chip::EndpointId endpoint, chip::MutableCharSpan & epUniqueIdMutSpan);
```
Again, weird to have an unconditional declaration that will then lead to link errors if CHIP_CONFIG_USE_ENDPOINT_UNIQUE_ID is false....

bzbarsky-apple commented on src/app/util/mock/MockNodeConfig.cpp:
```diff
@@ -186,10 +186,14 @@ const MockAttributeConfig * MockClusterConfig::attributeById(AttributeId attribu
                                        std::initializer_list<EmberAfDeviceType> aDeviceTypes,
                                        std::initializer_list<app::Clusters::Descriptor::Structs::SemanticTagStruct::Type> aTags,
-                                       app::EndpointComposition aComposition) :
+                                       app::EndpointComposition aComposition, chip::CharSpan aEndpointUniqueId) :
```
```suggestion
                                       app::EndpointComposition aComposition, CharSpan aEndpointUniqueId) :
```

bzbarsky-apple commented on src/app/util/mock/MockNodeConfig.h:
```diff
@@ -112,7 +112,8 @@ struct MockEndpointConfig
                        std::initializer_list<app::Clusters::Descriptor::Structs::SemanticTagStruct::Type> aTags = {},
-                       app::EndpointComposition composition = app::EndpointComposition::kFullFamily);
+                       app::EndpointComposition composition = app::EndpointComposition::kFullFamily,
+                       chip::CharSpan aEndpointUniqueID     = chip::CharSpan());
```
```suggestion
                       CharSpan aEndpointUniqueID     = CharSpan());
```

bzbarsky-apple commented on src/app/util/mock/MockNodeConfig.h:
```diff
@@ -133,6 +134,8 @@ struct MockEndpointConfig
     const EndpointId id;
     const app::EndpointComposition composition;
     const std::vector<MockClusterConfig> clusters;
+    char endpointUniqueIdBuffer[chip::app::Clusters::Descriptor::Attributes::EndpointUniqueID::TypeInfo::MaxLength()] = { 0 };
```
```suggestion
    char endpointUniqueIdBuffer[app::Clusters::Descriptor::Attributes::EndpointUniqueID::TypeInfo::MaxLength()] = { 0 };
```

bzbarsky-apple commented on src/python_testing/TC_DeviceBasicComposition.py:
I did not carefully review these changes.  But some of the renaming bits looks completely unrelated to this PR.

bzbarsky-apple commented on src/python_testing/matter_testing_infrastructure/chip/testing/matter_testing.py:
```diff
@@ -1142,27 +1136,6 @@ async def read_single_attribute(
         data = result[endpoint]
         return list(data.values())[0][attribute]
 
-    async def read_single_attribute_all_endpoints(
```
These changes seem rather unrelated to this PR....

bzbarsky-apple commented on src/python_testing/matter_testing_infrastructure/chip/testing/pics.py:
```diff
@@ -64,26 +64,17 @@ def parse_pics(lines: typing.List[str]) -> dict[str, bool]:
 
 
 def parse_pics_xml(contents: str) -> dict[str, bool]:
-    pics: dict[str, bool] = {}
```
This seems to be reverting other PRs.  Probably as a result of a bad merge?

Which makes me wonder what _else_ is being reverted, in the python_testing bits.  Please check these parts carefully.

suneelsignify replied:
It was a result of a bad merge, fixed now.

bzbarsky-apple reviewed (changes requested):
Changes requested due to the bad merge bits.

suneelsignify committed:
Complying to review feedback.

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

suneelsignify committed:
Fixing CI errors

suneelsignify committed:
Fixing CI errors

restyled-commits committed:
Restyled by autopep8

andy31415 commented on src/data-model-providers/codegen/tests/TestCodegenModelViaMocks.cpp:
```diff
@@ -2748,3 +2764,19 @@ TEST_F(TestCodegenModelViaMocks, ServerClusterInterfacesListClusters)
+    // Mock endpoint 4 has a unique ID
+    // ASSERT_TRUE(builder.IsEmpty()); // ownership taken above, we start fresh
+    ASSERT_EQ(model.EndpointUniqueID(kMockEndpoint4, span), CHIP_NO_ERROR);
+    EXPECT_TRUE(span.data_equal(CharSpan::fromCharString("AABBCCDDEEFFGGHHIIJJKKLLMMNNOO01")));
```
_span should be used instead of fromCharString

andy31415 commented on src/data-model-providers/codegen/tests/TestCodegenModelViaMocks.cpp:
```diff
@@ -2748,3 +2764,19 @@ TEST_F(TestCodegenModelViaMocks, ServerClusterInterfacesListClusters)
+    // ASSERT_TRUE(builder.IsEmpty()); // ownership taken above, we start fresh
+    ASSERT_EQ(model.EndpointUniqueID(kMockEndpoint4, span), CHIP_NO_ERROR);
+    EXPECT_TRUE(span.data_equal(CharSpan::fromCharString("AABBCCDDEEFFGGHHIIJJKKLLMMNNOO01")));
+    ASSERT_EQ(span.size(), strlen("AABBCCDDEEFFGGHHIIJJKKLLMMNNOO01"));
```
data_equal implies size matches.

andy31415 commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -187,7 +187,7 @@
 from chip.testing.taglist_and_topology_test import (create_device_type_list_for_root, create_device_type_lists,
                                                     find_tag_list_problems, find_tree_roots, flat_list_ok,
-                                                    get_direct_children_of_root, parts_list_problems, separate_endpoint_types)
+                                                    get_direct_children_of_root, parts_list_cycles, separate_endpoint_types)
```
PR descrition does not explain why anything under python_testing was changed.

These are unexpected changes and seem unrelated.

suneelsignify replied:
There were some merge issues, resolved.

andy31415 reviewed (dismissed):
Request changes: seeming unrelated changes in the python testing infrastructure.

suneelsignify committed:
Complying to review feedback

suneelsignify committed:
Fixed CI errors

suneelsignify committed:
Fixed CI errors

justinsignify494 committed:
Python Testing CI/CD test

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=37990) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=37990) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: suneelsignify<br/>:x: justinsignify494<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=37990) it.</sub>

justinsignify494 committed:
Python Testing CI/CD test fix 1.0

justinsignify494 committed:
Python Testing CI/CD test fix 1.1

justinsignify494 committed:
target return type fix as optional CI/CD fix 1.2

justinsignify494 committed:
ci/cd mypy validation fix

restyled-commits committed:
Restyled by autopep8

andy31415 commented on src/app/util/mock/attribute-storage.cpp:
```diff
@@ -244,6 +244,22 @@ chip::EndpointId emberAfEndpointFromIndex(uint16_t index)
+
+    // Copy the unique ID into the provided span
+    CharSpan targetSpan(epConfig->endpointUniqueIdBuffer, epConfig->endpointUniqueIdSize);
+    CopyCharSpanToMutableCharSpanWithTruncation(targetSpan, epUniqueIdSpan);
+
+    return CHIP_NO_ERROR;
+}
```
```suggestion
    return CopyCharSpanToMutableCharSpan(targetSpan, epUniqueIdSpan);
}
```

Seems more correct - why would we just truncate and silently ignore that we truncated?

andy31415 commented on src/python_testing/matter_testing_infrastructure/chip/testing/matter_testing.py:
```diff
@@ -291,6 +291,42 @@ class AttributeValue:
     timestamp_utc: Optional[datetime] = None
 
 
+class AttributeMatcher:
```
Please add to the description what changes are made and why. Right now it says "The Changes in the request are related to adding this new attribute." ... that is not enough: please specify in the summary on what changes were needed and why.

In particular I would like to understand why pics and matter testing code had to be changed for a new attribute ... that is unexpected.

suneelsignify replied:
Changes to pics.py and matter_testing.py files:

1. In test case TC-DESC-2.1 (https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/cluster/Descriptor.adoc) step 1a we need to refer PIXIT.DESC.DeviceTypeConformanceList In the first version we hardcoded it and the review feedback was to read it from the spec. The existing pics.py was not supporting to read PIXIT hence it had to be added. Following is the explanation how it is added :   
2. We added a new function `read_pixit_from_file` in pics.py file for reading PIXIT values from spec. Earlier there is no code to read PIXIT values from pics files, only picsItems was supported via function `read_pics_from_file`.
3. Above mentioned new function (`read_pixit_from_file` ) is called in matter_testing.py file which will return PIXIT values via  `matter_test_config` and can be used in test cases.

andy31415 reviewed (changes requested):
Please describe changes more than `The Changes in the request are related to adding this new attribute.`

Trying to understand what changes were made and why. It seems to be "add the new attribute in the XML" (that is obvious) then add ember storage for it (great ... who defines that and how does it get filled?) ... and then changes python? Why? what other things?

Please help me understand what to expect in the PR and what to consider "real changes" vs "accidental updates during merge"

suneelsignify committed:
Complying to review comment

suneelsignify committed:
Merge branch 'master' into master

suneelsignify committed:
Fixing CI issues

suneelsignify committed:
Fix for CI errors

suneelsignify committed:
Merge branch 'master' into master

suneelsignify committed:
Resolving CI errors

restyled-commits committed:
Restyled by prettier-yaml

suneelsignify committed:
Merge branch 'master' into master

suneelsignify committed:
Fixing the test cases

restyled-commits committed:
Restyled by autopep8

suneelsignify committed:
Update TC_DeviceBasicComposition.py

Fix CI errors

suneelsignify committed:
Merge branch 'master' into master

restyled-commits committed:
Restyled by isort

andy31415 commented on examples/bridge-app/linux/main.cpp:
```diff
@@ -257,8 +260,15 @@ DataVersion gComposedTempSensor2DataVersions[MATTER_ARRAY_SIZE(bridgedTempSensor
 
 // ---------------------------------------------------------------------------
 
+#if !CHIP_CONFIG_USE_ENDPOINT_UNIQUE_ID
 int AddDeviceEndpoint(Device * dev, EmberAfEndpointType * ep, const Span<const EmberAfDeviceType> & deviceTypeList,
                       const Span<DataVersion> & dataVersionStorage, chip::EndpointId parentEndpointId = chip::kInvalidEndpointId)
+#else
+
+int AddDeviceEndpoint(Device * dev, EmberAfEndpointType * ep, const Span<const EmberAfDeviceType> & deviceTypeList,
+                      const Span<DataVersion> & dataVersionStorage, chip::CharSpan epUniqueId,
+                      chip::EndpointId parentEndpointId = chip::kInvalidEndpointId)
+#endif
```
```suggestion
int AddDeviceEndpoint(Device * dev, EmberAfEndpointType * ep, const Span<const EmberAfDeviceType> & deviceTypeList,
                      const Span<DataVersion> & dataVersionStorage, 
#if !CHIP_CONFIG_USE_ENDPOINT_UNIQUE_ID                      
                      chip::CharSpan epUniqueId,
#endif                      
                      chip::EndpointId parentEndpointId = chip::kInvalidEndpointId)
```

Seems to make it more obvious about "what is different". 
It also changes a method with 5 parameters to 6 parameters, we may opt to provide some `Params` structure instead, in which case we would ifdef a member of that struct and it would be easier to read.

suneelsignify replied:
You meant 
`#if CHIP_CONFIG_USE_ENDPOINT_UNIQUE_ID                      
                      chip::CharSpan epUniqueId,
#endif `

andy31415 replied:
yes, copy and paste error. Will edit

andy31415 commented on src/app/util/attribute-storage.cpp:
```diff
@@ -336,6 +344,16 @@ CHIP_ERROR emberAfSetDynamicEndpoint(uint16_t index, EndpointId id, const EmberA
     emAfEndpoints[index].dataVersions   = dataVersionStorage.data();
+#if CHIP_CONFIG_USE_ENDPOINT_UNIQUE_ID
+    MutableCharSpan targetSpan(emAfEndpoints[index].endpointUniqueId);
+    CopyCharSpanToMutableCharSpanWithTruncation(endpointUniqueId, targetSpan);
```
why with truncation? Why not return error on failure and use the non-truncation version?

andy31415 commented on src/app/util/attribute-storage.cpp:
```diff
@@ -336,6 +344,16 @@ CHIP_ERROR emberAfSetDynamicEndpoint(uint16_t index, EndpointId id, const EmberA
+    static_assert(sizeof(emAfEndpoints[0].endpointUniqueId) <= UINT8_MAX,
+                  "The size of emAfEndpoints[index].endpointUniqueId must fit within uint8_t");
+
+    emAfEndpoints[index].endpointUniqueIdSize = static_cast<uint8_t>(targetSpan.size());
```
This is incorrect if we used truncation (if data was truncated,. the span content is smaller)

andy31415 commented on src/python_testing/matter_testing_infrastructure/chip/testing/matter_testing.py:
```diff
@@ -689,7 +689,6 @@ class MatterTestConfig:
     thread_operational_dataset: Optional[str] = None
 
     pics: dict[bool, str] = field(default_factory=dict)
-
```
revert changes to this file if it is a noop only.

andy31415 commented on src/python_testing/TC_DeviceBasicComposition.py:
```diff
@@ -941,6 +944,274 @@ def log_structured_data(start_tag: str, dump_string):
+                self.fail_current_test(
+                    "Device types not in ROOT_NODE_DEVICE_TYPE, POWER_SOURCE_DEVICE_TYPE, OTA_REQUESTOR_DEVICE_TYPE, OTA_PROVIDER_DEVICE_TYPE")
+
+        # TODO : Step 1b.4.1, 1b.4.2 needs to the run manually as PIXIT.DESC.DeviceTypeConformanceList
```
should we print-step here as "skipped" to notify that we are not performing it?
Please add an issue and reference it here so we can claim we recorded this as an issue of what needs to be done.

leorozendaal replied:
see https://csa-matter.slack.com/archives/C08Q1SXA9TL/p1745837714228939
we suggest to keep this at manual for TE2

suneelsignify replied:
Issue create : https://github.com/project-chip/connectedhomeip/issues/38640

suneelsignify committed:
Complying to review comments

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by autopep8

suneelsignify committed:
Merge branch 'master' into master

suneelsignify committed:
Zap generation for the new apps

suneelsignify committed:
complying to review

restyled-commits committed:
Restyled by clang-format
